controllers/bgppeer_controller.go

```diff
@@ -0,0 +1,146 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+	metallbv1alpha1 "github.com/metallb/metallb-operator/api/v1alpha1"
+	"github.com/metallb/metallb-operator/pkg/apply"
+	"github.com/metallb/metallb-operator/pkg/render"
+)
+
+// BGPPeer Reconciler reconciles a Peer object
+type BGPPeerReconciler struct {
+	client.Client
+	Log       logr.Logger
+	Scheme    *runtime.Scheme
+	Namespace string
+}
+
+const BGPPeerManifestPathContoller = "./bindata/configuration/bgp-peer"
+
+var BGPPeerManifestPath = BGPPeerManifestPathContoller
+
+//+kubebuilder:rbac:groups=metallb.io,resources=bgppeers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metallb.io,resources=bgppeers/status,verbs=get;update;patch
+
+func (r *BGPPeerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info(fmt.Sprintf("Starting BGP Peer reconcile loop for %v", req.NamespacedName))
+
+	instance := &metallbv1alpha1.BGPPeer{}
+	defer r.Log.Info(fmt.Sprintf("Finish BGP Peer reconcile loop for %v", req.NamespacedName))
+
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			err = r.syncMetalLBBGPPeers()
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	err := r.syncMetalLBBGPPeer(instance)
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("sync MetalLB BGP Peer failed %s", err))
+		return ctrl.Result{RequeueAfter: RetryPeriod}, err
```

Comment:
why requeueafter is needed here?

Reply:
I have been using flow so if the controller failed don't want to overwhelem the APIs with retries back to back hence the added delay ofcourse I read this somewhere or saw it in an example but don't recall where now if u want I can dig it out the same is also used in addresspool controller

Reply:
I think it makes sense. The only problem would be if the error is not transient (i.e. RenderObject fails or the configmap is malformed and the api server rejects it).
It would be good to intercept api timeouts errors vs "real errors", but the retry period is long enough that it won't harm (actually, I'd make it a bit shorter)

Reply:
ok dropped it to 1min